<a href="https://colab.research.google.com/github/j-claudinei-f/j-claudinei-f/blob/main/Inducao_e_Recorrencia.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Texto auxiliar sobre demonstrações por indução e relações de recorrência**

José Claudinei Ferreira

Universidade Federal de Alfenas

#**Uma infinidade de números primos**



Um número natural positivo que não pode ser escrito como produto de dois outros números naturais menores que ele é chamado de número primo.

**Teorema:** Existem inifnitos números primos.

**Demonstração:** Sabemos que um número natural $n$ é primo quando ele é divisível apenas por 1 e por ele mesmo.

Suponha que o teorema seja falso. Isso implicaria que o conjunto $P$ de todos os números primos seria finito, ou seja, teria $p$ elementos, sendo $p$ um número natural.

Então, poderíamos enumerar esses elementos de $P$ como $$P=\{n_1,\,n_2,\,\ldots,\, n_p\}.$$

Se esse fosse o caso, o número natural $$n_{p+1}=n_1\times n_2\times\cdots\times n_p+1$$ <font color=blue> não seria divisivel por nenhum $n_j\in P$ e, portanto, seria primo.

Isso geraria um problema, porque supomos que existiam apenas $p$ números primos e não $p+1$.

Logo, $P$ não pode ser finito.$\hspace{4cm}\square$

**Obs:** Não é sempre verdade que o número $n_{p+1}$ da demonstração é primo. Seria verdade se $P$ fosse finito, o que não é o caso. Mas se você testar alguns casos pode encontrar primos.

In [1]:
import sympy as sp # para usar a função isprime() que testa se um número é primo.

Testendo alguns casos:

In [ ]:
p=2*3+1
p,sp.isprime(p)

(7, True)

In [ ]:
p=2*3*5+1
p,sp.isprime(p)

(31, True)

In [ ]:
p=2*3*5*7+1
p,sp.isprime(p)

(211, True)

In [ ]:
p=2*3*5*7*11+1
p,sp.isprime(p)

(2311, True)

In [ ]:
p=2*3*5*7*11*13+1
p,sp.isprime(p)

(30031, False)

**Obs:** <font color=blue> Na verdade, na demonstração anterior apresentada usamos um resultado que podemos provar.

**Teorema 2:**

Um número natural $n$ ou é primo ou é o produto $n=n_1\times n_2$, sendo $n_1$ um número primo.

**Demonstração:** <font color=blue> Feito em sala.

**Corolário:** Um número natural $n$ ou é primo ou é o produto de números primos.

**Demonstração:** <font color=blue> Feito em sala.

**Corolário:** Um número natural $n$ ou é primo ou é o produto  $n=n_1\times n_2$, com $n_1$ primo e $1<n_1\leq \sqrt{n}$.

**Demonstração:** <font color=blue> Feito em sala, mas repito aqui. Do teorema temos que $n=n_1\times n_2$. Se $n_1>\sqrt{n}$ e $n_2>\sqrt{n}$, então $n_1\times n_2>\sqrt{n}\times \sqrt{n}=n$, o que não pode ocorrer.<br><br>
Logo, $n_1$ ou $n_2$ tem que ser menor ou igual a $\sqrt{n}$; e podemos escolher esse número menor primo.$\hspace{2cm}\square$

Esses resultados nos permitem verificar se um número é primo ou determinar seus fatores, como o algoritmo que segue:

In [ ]:
import math

def Fatores_primos(n):     # Só funciona para números naturais maiores que 1.
    while n % 2 == 0:      # imprime o número de vezes que 2 divide n.
        print(2)
        n = n // 2         # troca n por n/2 (divisão inteira)

    i=3
    while i*i<=n:          # testando números ímpares.
         while n % i== 0:  # imprime o número de vezes que i divide n.
            print(i)
            n = n // i     # troca n por n/i (divisão inteira)
         i=i+2

    if n > 2:
        print(n) # Se não hove divisão de $n$ ele é primo

**Obs:** Note que, caso prefira (ou possa) usar o cálculo de $\sqrt{n}$, você pode trocar o

```
while i*i<=n:
```
por

```
for i in range(3,int(math.sqrt(n))+1,2):
```

**Obs:** Uma curiosidade sobre o cálculo de $\sqrt{x}$ no [link](https://en.wikipedia.org/wiki/Fast_inverse_square_root).

Teste com $n=2\times3\times5\times7\times11\times13+1=30031$ que já testamos não ser primo.



In [ ]:
p=2*3*5*7*11*13+1
Fatores_primos(p)

59
509


**Obs:** <font color=blue> Note que $30031=59\times 509$ é produto de dois números primos, um menor e outro maior que sua raiz quadrada.

<font color=red> Isso nos diz que pode ocorrer um fator primo maior que a raiz quadrada, mas só pode ocorrer um.

Vamos testar com outro número:

In [ ]:
Fatores_primos(123456789)

3
3
3607
3803


Quando esse algoritmo retornar apenas o número $n$ de entrada, esse número será primo.

In [ ]:
Fatores_primos(2*3*5*7*11+1)

2311


#**A torre de Hanói**

No estudo da determinação do número mínimo de movimentos para mover os discos da [torre de Hanói](http://clubes.obmep.org.br/blog/torre-de-hanoi/) observamos que, se denotarmos por $p(n)$ o número mínimo de movimentos, para mover $n$ discos, com as **regras**:


1.   Mover um disco de cada vez.
2.   Não colocar disco maior sobre disco menor.

Teremos
$$p(1)=1,\qquad p(n+1)=2p(n)+1.$$

Como podemos determinar $p(6)$ e $p(10)$?





**Uma forma de resolver o problema, com o uso de um computador e programação:**

Podemos usar uma linguagem de programação (ou uma calculadora programável) para calcular $p(n)$. Nesse caso, vamos usar a linguagem Phyton.

In [ ]:
def p(n):         # O comando def é usado para definir funções das mais variadas.

   if (n==1):     # Se n=1, retorne p(1)=1
     q=1

   else:          # Senão, ou seja, se p for diferente de 1, retorne 2p(n-1)+1.
     q=2*p(n-1)+1 # Aqui a função é chamada outra vez, para m=n-1, e outra vez, para m=n-2, ..., até chegar em m=1.
                  # Após todas as chamadas retorna a m=n-1 e m=n.

   return q       # Essa é a saída.

**Obs.** Note que a função não está bem definida para números menores que 1, ou para números que não sejam naturais.

Vamos testar, para $n=1$:

In [ ]:
p(1)

1

Vamos testar, para $n=2$:

In [ ]:
p(2)

3

Vamos testar, para $n=3$:

In [ ]:
p(3)

7

Para ver vários valores de $p(n)$ de uma vez, por exemplo, $p(i)$, para $1\leq i\leq 10$:

In [ ]:
for i in range(1,10+1): # Para i variando de 1 até 10, escreva o par (i, p(i)).
  print(i,p(i))

1 1
2 3
3 7
4 15
5 31
6 63
7 127
8 255
9 511
10 1023


**Obs.** Voce poderia fazer isso de outra forma, sem usar a definição da função $p(i)$. Começando a enumerar todos os $p(i)$, a partir do 1, como fazemos a seguir:

In [ ]:
q=1                   # q=1 representa p(1)=1.
print(1,q)

for i in range(1,10): # Para i variando de 1 até 10, troque q por 2q+1, porque p(n)=2p(n-1)+1.
  q=2*q+1
  print(i+1,q)        # Escreva o par (i, 2p+1).

1 1
2 3
3 7
4 15
5 31
6 63
7 127
8 255
9 511
10 1023


Esse último procedimento é menos elegante, mas é muito mais rápido de ser executado.

**Obs.** Observe ainda que podemos demonstrar, usando [indução finita](https://pt.wikipedia.org/wiki/Indu%C3%A7%C3%A3o_matem%C3%A1tica), que $$p(n)=2^n-1.$$ Isso facilita bastante na resolução do problema, embora uma calculadora ajude muito ainda.

In [ ]:
for i in range(1,10+1): # Para i variando de 1 até 10.
    print(i,2**i-1)        # Escreva o par (i, 2^i-1).

1 1
2 3
3 7
4 15
5 31
6 63
7 127
8 255
9 511
10 1023


**Complicando um pouco mais a regra:**

No estudo da determinação do número mínimo de movimentos para mover os discos da [torre de Hanói](http://clubes.obmep.org.br/blog/torre-de-hanoi/) observamos que, se denotarmos por $p(n)$ o número mínimo de movimentos, para mover $n$ discos, com as **regras**:


1.   Mover um disco de cada vez.
2.   Não colocar disco maior sobre disco menor.
3.   Cada movimento de disco deve partir da haste central ou chegar nela.

Teremos
$$p(1)=2,\qquad p(n+1)=?.$$

Como podemos determinar $p(6)$ e $p(10)$?